File: crates/domain/src/reward.rs
//! Bits reward policy.
//!
//! The reward invariant from `docs/07-gamification.md` is:
//!
//! ```text
//! first-attempt success >= recovery after failure > skip/reveal (0)
//! ```
//!
//! Incorrect attempts never subtract Bits, and a retried question cannot earn the
//! full first-attempt reward again.

/// Base Bits for a first-attempt, fully-correct answer.
pub const BASE_BITS: i64 = 10;

/// Difficulty bonus in `[0, 6]`, derived from the question's difficulty prior.
pub fn difficulty_bonus(difficulty_prior: f64) -> i64 {
    (difficulty_prior.clamp(0.0, 1.0) * 6.0).round() as i64
}

/// Bits earned for one accepted attempt.
///
/// A non-fully-correct attempt earns nothing. The first fully-correct attempt
/// earns the full reward; a later fully-correct attempt (recovery) earns roughly
/// half. Values are deterministic so they are easy to test and audit.
pub fn reward_bits(attempt_number: i32, score: f64, difficulty_prior: f64) -> i64 {
    if score < 1.0 {
        return 0;
    }
    let full = BASE_BITS + difficulty_bonus(difficulty_prior);
    if attempt_number <= 1 {
        full
    } else {
        (full / 2).max(1)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn incorrect_attempts_earn_nothing() {
        assert_eq!(reward_bits(1, 0.0, 0.5), 0);
        assert_eq!(reward_bits(1, 0.99, 0.5), 0);
        assert_eq!(reward_bits(3, 0.5, 0.5), 0);
    }

    #[test]
    fn first_attempt_success_earns_full_reward() {
        assert_eq!(reward_bits(1, 1.0, 0.0), 10);
        assert_eq!(reward_bits(1, 1.0, 0.5), 13);
        assert_eq!(reward_bits(1, 1.0, 1.0), 16);
    }

    #[test]
    fn recovery_earns_less_than_first_attempt() {
        let first = reward_bits(1, 1.0, 0.6);
        let recovery = reward_bits(2, 1.0, 0.6);
        assert!(recovery > 0, "recovery must still be rewarded");
        assert!(recovery < first, "recovery must be less than first attempt");
        assert_eq!(recovery, (first / 2).max(1));
    }
}

File: apps/web/src/scoring/reward.ts
/**
 * Display-only Bits preview, mirroring the authoritative Rust reward policy.
 *
 * This exists purely for the immediate reward animation. It is never persisted
 * as settled currency: the wallet is settled server-side during `/v1/sync` and
 * reconciled from the returned balance. A learner cannot mint settled Bits by
 * editing this code.
 *
 * The preview can legitimately differ from the settled amount: the server
 * derives the attempt number from accepted history and uses the canonical
 * difficulty prior, while the browser only knows its local attempt count.
 */

/** Base Bits for a first-attempt, fully-correct answer. */
export const BASE_BITS = 10;

function clamp(value: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, value));
}

/** Difficulty bonus in `[0, 6]`, derived from the question's difficulty prior. */
export function difficultyBonus(difficultyPrior: number): number {
  return Math.round(clamp(difficultyPrior, 0, 1) * 6);
}

/**
 * Bits previewed for one locally scored attempt.
 *
 * Non-fully-correct attempts preview nothing; a later fully-correct attempt
 * previews roughly half. Settled Bits are always the server's decision.
 */
export function rewardBits(
  attemptNumber: number,
  score: number,
  difficultyPrior: number,
): number {
  if (score < 1) {
    return 0;
  }
  const full = BASE_BITS + difficultyBonus(difficultyPrior);
  return attemptNumber <= 1 ? full : Math.max(Math.floor(full / 2), 1);
}
